Filter access/egress stops for modes that involve driving

gmellemstrand opened the pull request:
### Summary
This PR adds a filter for access/egress searches with methods for filtering to the nearest stops only (number configureable) for modes that involve driving. The reason for this is discusses in the linked issue.

### Issue
Closes #3555

### Unit tests
Write a few words on how the new code is tested. 
- Were unit tests added/updated?
- Was any manual verification done?
- Any observations on changes to performance?
- Was the code designed so it is unit testable?
- Were any tests applied to the smallest appropriate unit?
- Do all tests pass [the continuous integration service](https://github.com/opentripplanner/OpenTripPlanner/blob/dev-2.x/docs/Developers-Guide.md#continuous-integration)?

### Code style
Have you followed the [suggested code style](https://github.com/opentripplanner/OpenTripPlanner/blob/dev-2.x/docs/Developers-Guide.md#code-style)? 

### Documentation
- Have you added documentation in code covering design and rationale behind the code?
- Were all non-trivial public classes and methods documented with Javadoc?
- Were any new configuration options added? If so were the tables in the [configuration documentation](https://github.com/opentripplanner/OpenTripPlanner/blob/dev-2.x/docs/Configuration.md) updated?

### Changelog
Was a bullet point added to the [changelog file](https://github.com/opentripplanner/OpenTripPlanner/blob/dev-2.x/docs/Changelog.md) with description and link to the linked issue?

leonardehrenfried commented:
This looks like an interesting solution to the problems that we have noticed, too.

We will definitely test this out but I cannot say when. Perhaps already next week.

leonardehrenfried commented on src/main/java/org/opentripplanner/routing/algorithm/raptor/router/street/AccessEgressFilter.java:
```diff
@@ -0,0 +1,59 @@
+  private static Collection<NearbyStop> filterCarPickup(
+      Collection<NearbyStop> nearbyStops, RoutingRequest request
+  ) {
+    return nearbyStops
```
I'm a little hesitant to be overly picky, but could this not be made a little more DRY by passing the number of stops as an argument to a function?

leonardehrenfried commented on src/main/java/org/opentripplanner/routing/algorithm/raptor/router/street/AccessEgressFilter.java:
```diff
@@ -0,0 +1,59 @@
+  ) {
+    return nearbyStops
+        .stream()
+        .sorted()
```
Out of curiosity: What is the sorting order here? Distance to the origin of the search?

gmellemstrand replied:
Yes, I can try to make it a bit more explicit

leonardehrenfried commented on src/main/java/org/opentripplanner/routing/algorithm/raptor/router/street/AccessEgressFilter.java:
```diff
@@ -0,0 +1,61 @@
+  /**
+   * Returns the closest stops by distance
+   */
+  private static Collection<NearbyStop> filterCarPickup(
```
Aren't `filterCarPickup` and `filterByCarPark` now identical and you only need one of them?

leonardehrenfried commented:
In the corresponding ticket you talk about filtering for the nearest car parking lots but I cannot see anything about this in the code. Is this intentional?

gmellemstrand commented:
The only reason is that I haven't gotten around to it yet.

gmellemstrand commented:
I have added some logic to the two filters now.

For the `filterCarPickup` method, stops that are using the same station as stops within range are also added.

For the `filterByCarPark` method, stops that use the same car park as stops within range are also added.

This is what I have come up with so far, but we might want to consider other approaches.

abyrd commented:
Gard and I just discussed this at length during a call. Because it is probably quite complex to solve the true underlying problem described in linked issue #3555 (over-branching due to small trade-offs between time and cost, which translate to trade-offs between driving and transit, and difficulty of defining "acceptable" trade-offs to consider) simply limiting the number of stops reached is intended as first attempt, just to get some measure in place and keep search times reasonable.

Gard indicated though that he's not really happy with the current logic, and liked another idea of only retaining stops within max(distance_to_closest_stop * expansion_factor, min_search_radius). The current system already has swappable strategies for filtering access stops, so this new strategy could be tried out alongside the existing one before making a final decision.

leonardehrenfried commented:
We are also interested in testing and providing feedback for this feature.

gmellemstrand commented:
After discussing this some more with @abyrd , I tried another approach. Now it limits the access and egress results to the closest stations per pattern, up to a certain number (currently 5). I have tested with higher limits, and it looks like it is possible to set it significantly higher without affecting performance too much. If we do that, however, we will need to add a new ItineraryFilter, so that we do not get unwanted results with long driving times.

The new filter looks promising, and I do not think there is too much of a chance that we filter out results that we actually want.

leonardehrenfried commented on src/main/java/org/opentripplanner/routing/algorithm/raptor/router/street/AccessEgressFilter.java:
```diff
@@ -29,16 +34,61 @@
+        .stream()
+        .sorted()
+        .filter(s -> s.stop instanceof Stop)
+        .map(s -> ((Stop) s.stop).getParentStation())
```
In our feed we don't have parent stations so they are null. This leads to a list with a single null value being returned here.

If there is no parent station then simply the stop itself ought to be user, isn't it?

gmellemstrand replied:
I didn't really think of that, as I am so used to stops always having a station.

We probably should have the code handle this, but I think the code here and in some other places too, would be a lot cleaner if we knew all the stops had stations (maybe by adding stations for all stops in the gtfs import process).

gmellemstrand replied:
We should also think a bit about how it should be mixed, because finding the 5 closest stops can be very different from finding the 5 closest stations.

hannesj commented:
I removed Entur test label, instead we should enable VehicleToStopHeuristics in otp config

leonardehrenfried commented:
Replaced by #3906
